NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.cpp

```diff
@@ -0,0 +1,173 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "LocalLinearLeastSquaresExtrapolator.h"
+
+#include <functional>
+
+#include <Eigen/Core>
+#include <logog/include/logog.hpp>
+
+#include "MathLib/LinAlg/LinAlg.h"
+#include "MathLib/LinAlg/MatrixVectorTraits.h"
+#include "NumLib/Assembler/SerialExecutor.h"
+#include "NumLib/Function/Interpolation.h"
+#include "ExtrapolatableElementCollection.h"
+
+namespace NumLib
+{
+LocalLinearLeastSquaresExtrapolator::LocalLinearLeastSquaresExtrapolator(
+    NumLib::LocalToGlobalIndexMap const& dof_table)
+    : _nodal_values(NumLib::GlobalVectorProvider::provider.getVector(
+          MathLib::MatrixSpecifications(dof_table.dofSizeWithoutGhosts(),
+                                        dof_table.dofSizeWithoutGhosts(),
+                                        &dof_table.getGhostIndices(),
+                                        nullptr)))
+#ifndef USE_PETSC
+    , _residuals(dof_table.size())
+#else
+    , _residuals(dof_table.size(), false)
+#endif
+    , _local_to_global(dof_table)
+{
+    /* Note in case the following assertion fails:
+     * If you copied the extrapolation code, for your processes from
+     * somewhere, note that the code from the groundwater flow process might
+     * not suit your needs: It is a special case and is therefore most
+     * likely too simplistic. You better adapt the extrapolation code from
+     * some more advanced process, like the TES process.
+     */
+    assert(dof_table.getNumberOfComponents() == 1 &&
+           "The d.o.f. table passed must be for one variable that has "
+           "only one component!");
+}
+
+void LocalLinearLeastSquaresExtrapolator::extrapolate(
+    ExtrapolatableElementCollection const& extrapolatables)
+{
+    _nodal_values.setZero();
+
+    // counts the writes to each nodal value, i.e., the summands in order to
+    // compute the average afterwards
+    auto counts =
+        MathLib::MatrixVectorTraits<GlobalVector>::newInstance(_nodal_values);
+    counts->setZero();  // TODO BLAS?
```

Comment:
The todo comment could be more specific, like "TODO: setZero might be implemented in the matrix interface."